From d223a00cb2751727b240728c0f32c82012fa8ed3 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Thu, 16 Jun 2022 12:52:24 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: nic reset after bist

In order to avoid ECC after running BIST on NIC rings,
support added to perform NIC unit reset.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: I2ca6014b5951f05d985abb65a9c3c91a97ed749f
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index 56eeb41..cffe9e1 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -1216,16 +1216,17 @@
 
 /* TODO: Add Xbar memory init after bist to avoid ecc. */
 /* TODO: Add vdec memory init after bist to avoid ecc. */
-/* TODO: Add NIC unit reset after bist to avoid ecc. */
 
 static void gaudi2_reset_engines(void)
 {
-	u32 tpc_iso_val, hbm_iso_val;
+	u32 tpc_iso_val, hbm_iso_val, nic_iso_val;
 
 	tpc_iso_val = regmap_readl(mmPSOC_GLOBAL_CONF_TPC_ISO);
 	regmap_writel(0, mmPSOC_GLOBAL_CONF_TPC_ISO);
 	hbm_iso_val = regmap_readl(mmPSOC_GLOBAL_CONF_HBM_ISO);
 	regmap_writel(0, mmPSOC_GLOBAL_CONF_HBM_ISO);
+	nic_iso_val = regmap_readl(mmPSOC_GLOBAL_CONF_NIC_ISO);
+	regmap_writel(0, mmPSOC_GLOBAL_CONF_NIC_ISO);
 
 	regmap_writel(PSOC_RESET_CONF_TPC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_TPC_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_TPC_DIV_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_TPC_DIV_UNIT_RST);
@@ -1233,6 +1234,8 @@
 	regmap_writel(PSOC_RESET_CONF_SM_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_SM_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_HBM_MC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_HBM_MC_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_MME_EUS_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_MME_EUS_UNIT_RST);
+	regmap_writel(PSOC_RESET_CONF_NIC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_NIC_UNIT_RST);
+	regmap_writel(PSOC_RESET_CONF_NIC_PRT_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_NIC_PRT_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_UNIT_RST_N_IND_MASK, mmPSOC_RESET_CONF_UNIT_RST_N);
 	udelay(DELAY_1K_USEC); /* TODO: confirm the delay */
 	regmap_writel(0, mmPSOC_RESET_CONF_UNIT_RST_N);
@@ -1242,11 +1245,14 @@
 	regmap_writel(0, mmPSOC_RESET_CONF_ROTATOR_UNIT_RST);
 	regmap_writel(0, mmPSOC_RESET_CONF_HBM_MC_UNIT_RST);
 	regmap_writel(0, mmPSOC_RESET_CONF_MME_EUS_UNIT_RST);
+	regmap_writel(0, mmPSOC_RESET_CONF_NIC_UNIT_RST);
+	regmap_writel(0, mmPSOC_RESET_CONF_NIC_PRT_UNIT_RST);
 	udelay(DELAY_1K_USEC); /* TODO: confirm the delay */
 
 	/* Restore original iso value */
 	regmap_writel(tpc_iso_val, mmPSOC_GLOBAL_CONF_TPC_ISO);
 	regmap_writel(hbm_iso_val, mmPSOC_GLOBAL_CONF_HBM_ISO);
+	regmap_writel(nic_iso_val, mmPSOC_GLOBAL_CONF_NIC_ISO);
 }
 
 static void init_sram_bist_data(void)
